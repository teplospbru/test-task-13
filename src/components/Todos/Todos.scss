@import '../../scss/variables';
@import '../../scss/fonts';

.todos {
    flex: auto;
    overflow-y: auto;

    @media  screen and (max-width: 768px) {
        max-height: 100%;
    }
}

.label {
    display: block;
    width: 100%;
    padding: 1.2rem 1rem 1.2rem 110px;

    @media  screen and (max-width: 1024px) {
        padding-left: 74px;
    }
    border-bottom: 1px solid $light-color;
    @include font(32, 300, 1.3);   
    color: $primary-color;
    position: relative;
    cursor: pointer;
    input[type="checkbox"] {
        appearance: none;
    }
    .checkbox__checked, .checkbox__unchecked {
        display: block;
        margin: 0 31px;
        width: 48px;
        height: 48px;
        position: absolute;
        top: 50%;
        left: 0;
        transform: translateY(-50%);

        @media  screen and (max-width: 1024px) {
            margin: 0 21px;
            width: 32px;
            height: 32px;
        }
        svg {
            stroke: $light-color;
        }
    }
    .checkbox__checked {
        display: none;
        svg {
            stroke: $green-color;
        }
    }
    input[type='checkbox']:checked ~ .checkbox__checked {
        display: block;
    }
    span ~ input[type='checkbox']:checked {
        display: block;
        color: $primary-color;
    }

    @media  screen and (max-width: 1024px) {
        @include font(24, 300, 1.3);
    }
}