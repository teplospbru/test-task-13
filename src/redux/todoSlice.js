import { createSlice } from "@reduxjs/toolkit";

export const todoSlice = createSlice({
    name: 'todos',
    initialState: {
        todos: [
            { id: 1, body: 'Next, we need to import the reducer function from the counter slice and add it to our store', isCompleted: false },
            { id: 2, body: 'Creating a slice requires a string name to identify the slice', isCompleted: true },
            { id: 3, body: 'Provide the Redux Store to React', isCompleted: false },
        ],
    },
    reducers: {
        createTodo(state, action) {
            state.todos.push({ id: Date.now(), body: action.payload, isCompleted: false })
        },
        setCompleted(state, action) {
            const index = state.todos.findIndex(todo => todo.id == action.payload);
            state.todos[index].isCompleted = !state.todos[index].isCompleted;
        },
        clearCompleted(state, action) {
            state.todos = state.todos.filter(todo => todo.isCompleted === false);
        }
    }
});

export const { clearCompleted, createTodo, setCompleted } = todoSlice.actions;

export default todoSlice.reducer;