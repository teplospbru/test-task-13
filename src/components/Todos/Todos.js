import './Todos.scss';
import { useDispatch } from 'react-redux';
import { setCompleted } from '../../redux/todoSlice';
import { useEffect, useState, useRef } from 'react';

const styles = {
    textDecoration: 'line-through',
    fontStyle: 'italic',
    color: '#d1d1d1'
}

const Todos = ({ todos }) => {
    const dispatch = useDispatch();
    const div = useRef();
    const [height, setHeight] = useState('');

    useEffect(() => {
        setHeight(div.current.getBoundingClientRect().height + 'px');
    }, [div]);

    const checkboxHandler = (id) => {
        dispatch(setCompleted(id));
    }
    
    return (
        <div className='todos' ref={ div } style={{ height }}>
            {
                todos.map(({ id, body, isCompleted }) => (
                    <label className="label" htmlFor={ id } key={ id + '_1' }>
                        <input 
                            type="checkbox" 
                            name="policy" id={ id } 
                            checked={ isCompleted } 
                            onChange={ () => checkboxHandler(id) }
                            
                        ></input>
                        <span style={ isCompleted ? styles : null }>{ body }</span>
                        <div className="checkbox__unchecked">
                            <svg>
                                <use xlinkHref="#unchecked"></use>
                            </svg>
                        </div>
                        <div className="checkbox__checked">
                            <svg>
                                <use xlinkHref="#checked"></use>
                            </svg>
                        </div>
                    </label>
                ))
            }
        </div>
    )
};

export default Todos;