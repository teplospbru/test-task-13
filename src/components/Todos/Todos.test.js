import Todos from './Todos';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../redux/store';
import renderer from 'react-test-renderer';

const TodoArr = [
    { id: 1, body: 'Next, we need to import the reducer function from the counter slice and add it to our store', isCompleted: false },
    { id: 2, body: 'Creating a slice requires a string name to identify the slice', isCompleted: true },
    { id: 3, body: 'Provide the Redux Store to React', isCompleted: false },
];

beforeEach(() => {
    render(
        <Provider store={ store }>
            <Todos todos={ TodoArr } />
        </Provider>
    )
})

describe('testing the component "Todos"', () => {
    it('renders "Todos"', () => {
        const todos = screen.getByTestId('todos');

        expect(todos).toBeInTheDocument();
    });
    it('renders 3 todo items', () => {
        const todos = screen.getAllByTestId('todo');

        expect(todos.length).toBe(3);
    });
    it('renders 1 completed and 2 uncompleted todo items', () => {
        const todos = screen.getAllByTestId('span');

        expect(todos[0]).not.toHaveStyle('text-decoration: line-through');
        expect(todos[1]).toHaveStyle('text-decoration: line-through');
        expect(todos[2]).not.toHaveStyle('text-decoration: line-through');
    });
    it('matches to snapshot', () => {
        const container = renderer.create(<Provider store={ store }><Todos todos={ TodoArr }/></Provider>).toJSON();

        expect(container).toMatchSnapshot()
    });
});