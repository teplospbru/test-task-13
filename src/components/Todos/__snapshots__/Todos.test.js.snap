// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the component "Todos" matches to snapshot 1`] = `
<div
  className="todos"
  data-testid="todos"
  style={
    Object {
      "height": "",
    }
  }
>
  <label
    className="label"
    data-testid="todo"
    htmlFor={1}
  >
    <input
      checked={false}
      id={1}
      name="policy"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      data-testid="span"
      style={null}
    >
      Next, we need to import the reducer function from the counter slice and add it to our store
    </span>
    <div
      className="checkbox__unchecked"
    >
      <svg>
        <use
          xlinkHref="#unchecked"
        />
      </svg>
    </div>
    <div
      className="checkbox__checked"
    >
      <svg>
        <use
          xlinkHref="#checked"
        />
      </svg>
    </div>
  </label>
  <label
    className="label"
    data-testid="todo"
    htmlFor={2}
  >
    <input
      checked={true}
      id={2}
      name="policy"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      data-testid="span"
      style={
        Object {
          "color": "#d1d1d1",
          "fontStyle": "italic",
          "textDecoration": "line-through",
        }
      }
    >
      Creating a slice requires a string name to identify the slice
    </span>
    <div
      className="checkbox__unchecked"
    >
      <svg>
        <use
          xlinkHref="#unchecked"
        />
      </svg>
    </div>
    <div
      className="checkbox__checked"
    >
      <svg>
        <use
          xlinkHref="#checked"
        />
      </svg>
    </div>
  </label>
  <label
    className="label"
    data-testid="todo"
    htmlFor={3}
  >
    <input
      checked={false}
      id={3}
      name="policy"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      data-testid="span"
      style={null}
    >
      Provide the Redux Store to React
    </span>
    <div
      className="checkbox__unchecked"
    >
      <svg>
        <use
          xlinkHref="#unchecked"
        />
      </svg>
    </div>
    <div
      className="checkbox__checked"
    >
      <svg>
        <use
          xlinkHref="#checked"
        />
      </svg>
    </div>
  </label>
</div>
`;
